File: user/src/main/scala/ru/tdd/app/models/exceptions/ApiException.scala
package ru.tdd.app.models.exceptions

import zio.http.Status
import zio.http.Status.{BadRequest, Conflict, NotFound, UnprocessableEntity}

import java.time.LocalDateTime

/**
 * @author Tribushko Danil
 * @since 05.11.2025
 * @param statusCode http код ошибки
 * @param message    текст сообщения
 * @param timestamp  время ошибки
 *                   Общее исключение api приложения
 */
class ApiException(
                    val statusCode: Status,
                    val message: String,
                    val timestamp: LocalDateTime = NOW_TIME
                  ) extends RuntimeException(message)

/**
 * @author Tribushko Danil
 * @since 05.11.2025
 */
case class BadRequestException(
                                override val message: String,
                                override val timestamp: LocalDateTime = NOW_TIME
                              ) extends ApiException(BadRequest, message, timestamp)

/**
 * @author Tribushko Danil
 * @since 05.11.2025
 */
case class ValidationException(
                                override val message: String,
                                override val timestamp: LocalDateTime = NOW_TIME
                              ) extends ApiException(UnprocessableEntity, message, timestamp)

/**
 * @author Tribushko Danil
 * @since 05.11.2025
 */
case class NotFoundException(
                              override val message: String,
                              override val timestamp: LocalDateTime = NOW_TIME
                            ) extends ApiException(NotFound, message, timestamp)

/**
 * @author Tribushko Danil
 * @since 05.11.2025
 */
case class AlreadyExistException(
                                  override val message: String,
                                  override val timestamp: LocalDateTime = NOW_TIME
                                ) extends ApiException(Conflict, message, timestamp)

File: user/src/main/java/ru/tdd/user/application/models/exceptions/ApiException.java
package ru.tdd.user.application.models.exceptions;

import java.time.LocalDateTime;

/**
 * @author Tribushko Danil
 * @since 01.02.2026
 * Класс-исключения api приложения
 */
public class ApiException extends RuntimeException {

    protected final int statusCode;

    protected final LocalDateTime timestamp;

    public ApiException(int statusCode, String message) {
        super(message);
        this.statusCode = statusCode;
        this.timestamp = LocalDateTime.now();
    }

    public int getStatusCode() {
        return statusCode;
    }

    public LocalDateTime getTimestamp() {
        return timestamp;
    }
}
